apps/shared/avifpng.c

```diff
@@ -196,6 +198,70 @@ static avifBool avifExtractExifAndXMP(png_structp png, png_infop info, avifBool
     return AVIF_TRUE;
 }
 
+// Converts a gamma-compressed value to linear.
+static double avifToLinear(double v, double gamma)
+{
+    return pow(v, gamma);
+}
+
+// Gamma-compresses a linear value using the sRGB transfer curve.
+// See section 8 of https://www.color.org/srgb.pdf
+#define SRGB_GAMMA_SLOPE 12.92
+#define SRGB_GAMMA_TRESHOLD 0.0031308
+static double avifToSrgbGamma(double linear)
+{
+    if (linear <= SRGB_GAMMA_TRESHOLD) {
+        return linear * SRGB_GAMMA_SLOPE;
+    } else {
+        return pow(linear, 1. / 2.4) * 1.055 - 0.055;
+    }
+}
+
+static int avifCorrectGamma(uint16_t v, double gamma, int rgbMaxChannel)
+{
+    const double normalized = v / (double)rgbMaxChannel;
+    const double corrected = avifToSrgbGamma(avifToLinear(normalized, gamma));
+    const int scaledUp = (int)floor(corrected * rgbMaxChannel + 0.5);
+    return scaledUp < 0 ? 0 : scaledUp > rgbMaxChannel ? rgbMaxChannel : scaledUp;
```

Comment:
In theory there should be no need for clamping, but I feel like it's safer in case there's a precision error or something? But I can remove the clamping if you prefer.